### PR TITLE
8287555: Tighten G1 G1BlockOffsetTable::block_start() code

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  can I have reviews for this change that specializes `G1BlockOffsetTablePart::block_size` with a variant used for card aligned addresses including some cleanup of comments? This saves some unnecessary lookup of the block size of the object found as for refinement the start address is always aligned.

Testing: tier1, gha

Thanks,
  Thomas

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8287555](https://bugs.openjdk.org/browse/JDK-8287555): Tighten G1 G1BlockOffsetTable::block_start() code

### Reviewers
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9059/head:pull/9059` \
`$ git checkout pull/9059`

Update a local copy of the PR: \
`$ git checkout pull/9059` \
`$ git pull https://git.openjdk.org/jdk pull/9059/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9059`

View PR using the GUI difftool: \
`$ git pr show -t 9059`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9059.diff">https://git.openjdk.org/jdk/pull/9059.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp:166]: I am not sure introducing another API is really that necessary. What I have in mind is sth like:

```c
void block_start(const void* addr) {
  if (card_aligned(addr)) {
    fast_path
  } else {
    slow_path
  }
  assert_post-cond
}
```

I don't think the cost of if-check will be observable.

Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp:166]: I had something like that, but the "not observable" part was not that clear in my results (i.e. during garbage collection) so I only optimized the part where this can be statically determined for now.

Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp:166]: CPU resources like branch prediction/speculative execution aren't free either, and this is a very performance sensitive area of code.

Comment (kimbarrett) [reply src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp:166]: Then why not just card-align addr and always use the fast path code?

Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp:166]: @kimbarrett: The `block_start` method returns the first object reaching into `addr`; if we aligned first, then the result would be the first object reaching into `aligned(addr)` which is different; the callers would need to advance forward not-requested objects then.
Or maybe I'm missing something in your suggestion?

Comment (albertnetymk) [reply src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp:166]: > I had something like that, but the "not observable" part was not that clear in my results

I added the following to `G1BlockOffsetTablePart::block_start` to see the cost.

```c
  if (is_aligned(addr, BOTConstants::card_size())) {
    return q;
  }
```

Using `Scan Heap Roots (ms)` and `Scanned Cards` from `-Xlog:gc,gc+phases`, I can calculate #cards processed per ms. Surprisingly, I can observe some reduction (~2.7%) in this metric on my box.

Re Kim's suggestion, the model where `G1BlockOffsetTablePart::block_start` accepts only card-aligned addresses and the caller (`HeapRegion::block_start`) handles block-walking (heap parsing) makes more sense to me. (After all, "block size" is sth meaningful only inside `HeapRegion`.) Ofc, this can be dealt with in this PR or another.

Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp:166]: [Here](https://github.com/openjdk/jdk/compare/master...tschatzl:jdk:submit/move-block-size-out-of-bot?expand=1)'s a prototype for (trying to) move `block_start` out of `G1BlockOffsetTablePart`; in the end I have a bit of mixed feelings about it:
* the BlockOffsetTable deals with blocks; blocks intrinsically have a start and a size; removing the BlockOffsetTable to only deal with parts of a block seems a bit limited.
Additionally not knowing about size will require to move out `verify` too if we are strict about this distinction, which means that lots of code is being dumped in the already (too) large `HeapRegion` class. Of course, like in the prototype, the `G1BlockOffsetTablePart::verify` could also stay as is....
* the BlockOffsetTable is already very tightly dependent on `HeapRegion` anyway (for current layout, i.e. `bottom`, `top`, `end`).

Comments?

**Comment by tschatzl** [comment]
Did some refactoring to remove some duplicate `block_start` call. Passes tier1-3.

**Comment by tschatzl** [comment]
I moved some fairly unrelated refactoring/cleanup to separate PRs without updating this one yet (https://github.com/openjdk/jdk/pull/9430, https://github.com/openjdk/jdk/pull/9429 and https://github.com/openjdk/jdk/pull/9428); let's wait until they are pushed and merged.

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Comment by tschatzl** [comment]
It has been determined that this is too much code complexity for this non-observable gain. Closing for now.

**Pull request closed without merging**
